armbian/base/scripts/autosetup-ssd.sh

```diff
@@ -0,0 +1,209 @@
+    printf "%-16s %-7s %16s   %-30s\n" "---------------" "------" "-----------------" "------------------------------"
+
+    device_found=0
+    targets_total=0
+    blockdev_target_path=""
+
+    # loop over all top-level block devices
+    while read blockdev; do
+        name=$( echo "${blockdev}" | cut -f 1 -d " " )
+        type=$( echo "${blockdev}" | cut -f 2 -d " " )
+        size=$( echo "${blockdev}" | cut -f 4 -d " " )
+        blockdev_target=""
+
+        # check for existing filesystem
+        if [[ ${#type} -gt 0 ]]; then
+            blockdev_target="NO: has file system"
+
+        # check if at least 200GB
+        elif [[ ${size} -lt 200000000000 ]]; then
+            blockdev_target="NO: too small (min 200GB)"
+
+        else
+            # check top-level device for partitions
+            partition_count=0
+            while read partition; do
+                name_part=$( echo "${partition}" | cut -f 1 -d " " )
+                if [[ "${name_part}" =~ "${name}" ]] && [[ "${name_part}" != "${name}" ]]; then
+                    partition_count=$((partition_count+1))
+                    blockdev_target="NO: has ${partition_count} partition(s)"
+                fi
+            done <<< "$(lsblk -o NAME,FSTYPE,PARTTYPE,SIZE,TYPE,MAJ:MIN -abrnp -e 1,7,31,179,252)"
+
+            if [[ ${#blockdev_target} -eq 0 ]]; then
+                targets_total=$((targets_total + 1))
+                blockdev_target="OK"
+                blockdev_target_path="${name}"
+            fi
+
+            # check if specified device is found
+            if [[ "${name}" == "${DEVICE}" ]]; then
+                device_found=1
+            fi
+
+        fi
+
+        printf "%-16s %-7s %16s    %-20s\n" "${name}" "${type}" "${size}" "${blockdev_target}"
+
+    # `lsblk` returns only top-level devices (-d) and excludes device-types by major number
+    # (see http://www.lanana.org/docs/device-list/devices-2.6+.txt)
+    done <<< "$(lsblk -o NAME,FSTYPE,PARTTYPE,SIZE,TYPE,MAJ:MIN -abrnp -e 1,7,31,179,252 -d)"
+
+    printf "%-16s %-7s %16s   %-20s\n" "---------------" "------" "-----------------" "------------------------------"
+    printf "TOTAL %s potential target blockdevices found\n\n" "${targets_total}"
+}
+
+
+function format() {
+    ### DANGER ZONE ###
+    (
+        echo o # Create a new empty DOS partition table
+        echo n # Add a new partition
+        echo p # Primary partition
+        echo 1 # Partition number
+        echo   # First sector (Accept default: 1)
+        echo   # Last sector (Accept default: varies)
+        echo w # Write changes
+    ) | fdisk ${DEVICE}
+
+    case ${DEVICE} in
+        # internal drive (e.g. PCIe)
+        /dev/nvme*)
+            PARTITION="${DEVICE}p1"
+            ;;
+        # external drive (e.g. USB)
+        /dev/sd*)
+            PARTITION="${DEVICE}1"
+            ;;
+    esac
+
+    partprobe ${PARTITION}
```

Comment:
I found in manual testing that there seems to be a race condition here; `partprobe ${PARTITION}` can sometimes fail, seemingly due to the `fdisk` command existing before the new partition exists:
```
root@bitbox-base:~# bash autosetup-ssd.sh format auto --assume-yes
[...]
The partition table has been altered.
Calling ioctl() to re-read partition table.
Syncing disks.

Error: Could not stat device /dev/nvme0n1p1 - No such file or directory.
```

Changing the line to `partprobe ${DEVICE}` however always seems to succeed, since the device file will exist regardless.

Reply:
done